### PR TITLE
[autoscaler][kubernetes] Add ability to not copy cluster config to head node when calling `create_or_update_head_node`.

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

In situations where we don't run the monitor on the head node, there's no point in copying the cluster launching config onto the head node. 
The copying operation has lead to issues with the K8s operator (https://github.com/ray-project/ray/issues/13569).
 
This PR makes the following changes:

- Adds a `no_monitor_on_head_node` boolean argument to `commands.create_or_update_or_cluster`. 
When set to True, the cluster launching config is not copied to the head node.  

- Updates the K8s operator is to use the new flag.

- Updates the K8s operator example configs  to use the `--no-monitor` flag (https://github.com/ray-project/ray/pull/13505) in the head ray start commands. 

- To clean up the code a bit, I hid the config-copying logic behind a function call.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(
   
  Tested manually by running the unit tests for K8s operator and Ray on K8s cluster launcher.

### CODE REVIEW

File: python/ray/autoscaler/_private/commands.py
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/commands.py:664]: You do not need the ssh private key?

File: python/ray/autoscaler/_private/commands.py
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/commands.py:716]: can you add type annotation and function level description?
Any idea for a better name since we already have provider.prepare_for_head_node?

File: python/ray/autoscaler/_private/commands.py
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/commands.py:723]: you don't need the config["auto"] updates?

File: python/ray/autoscaler/_private/commands.py
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/commands.py:734]: you don't need "no_restart"?

File: python/ray/autoscaler/_private/commands.py
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/commands.py:140]: do you need to also update the sdk `create_or_update_cluster` ?

Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/commands.py:140]: hmm, sure I could, at least for the sake of consistency add `no_monitor_on_head: bool = False` to the `commands.py` function call inside of `sdk.create_or_update_cluster`

Probably don't want to add this yet to the arguments of `sdk.create_or_update_cluster`

The operator currently uses the `commands.create_or_update_cluster`

File: python/ray/autoscaler/_private/commands.py
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/commands.py:734]: well... at least not for the kubernetes operator

Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/commands.py:734]: For the K8s operator, none of the `remote_config` logic is needed.

**Comment by DmitriGekhtman** [comment]
Hmm, yyeah, I guess these changes are useful only for the Kubernetes operator.

File: python/ray/autoscaler/_private/commands.py
Comment (dmatch01) [inline python/ray/autoscaler/_private/commands.py:498]: Just wanted to clarify the expected logic for this line in the K8s operator use case where I assume `no_monitor_on_head`=`true`:

The warning will be logged.  Is that the correct/intended interpretation?

**Comment by DmitriGekhtman** [comment]
If no_monitor_on_head is true, the if condition becomes
not (A or True) = False,
so the warning isn’t logged.

The readability here could be improved.

File: python/ray/autoscaler/_private/commands.py
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/commands.py:498]: If no_monitor_on_head is true, the if condition becomes
not (A or True) = False,
so the warning isn’t logged.

The readability here could be improved.

Comment (dmatch01) [reply python/ray/autoscaler/_private/commands.py]: Got it.  And I see you improved read ability as well in the PR update.  Thank you!

**Comment by AmeerHajAli** [comment]
This looks good to me. Can we add a test that verifies it solves the issue?

**Comment by DmitriGekhtman** [comment]
> This looks good to me. Can we add a test that verifies it solves the issue?

Added a test which confirms that no files are mounted during the operator's Ray cluster creation.

**Comment by AmeerHajAli** [comment]
The test you added fails for any file mounts. Does that mean that the operator users should have nothing in their file_mounts config? Why not throw an error if they explicitly tried to mount?

**Comment by DmitriGekhtman** [comment]
Good question -- there's actually no user-interface for specifying file
mounts for the operator.

Users configure and apply a Kubernetes custom resource, which is internally
translated to an autoscaling config.
There's no file mounts field for the custom resource and the internal
autoscaling config has no file mounts.

This does suggest an improvement to the test: instead of reading in an
autoscaling config, read in a cluster custom resource and translate it to
an autoscaling config as the operator does.
This has the bonus that it tests the function that does the translation.

Will go ahead and make this change in the test.

**Comment by DmitriGekhtman** [comment]
Will wait for tests to finish.

**Comment by DmitriGekhtman** [comment]
Screen Shot 2021-02-01 at 12 33 37 PM
@AmeerHajAli 
Tests look good, pending final review.

**Comment by DmitriGekhtman** [comment]
Added a comment describing the additional flag.

File: python/ray/autoscaler/_private/commands.py
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/commands.py:724]: Isn't this:
`Tuple[Dict[str, Any], Dict[str, Any]]`?
Can we define a global variable `ClusterConfig = Dict[str, Any]` and use it in the 3 places?

**Review by AmeerHajAli — APPROVED** [review]
LGTM. left a minor comment.

File: python/ray/autoscaler/_private/commands.py
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/commands.py:724]: The second return item is a named temporary file (note doc string). Unfortunately, it's hard to type those due to the structure of the `tempfile` module (https://stackoverflow.com/questions/64429113/how-should-a-namedtemporaryfile-be-annotated)

I think `ClusterConfig = Dict[str, Any]` would be a good thing to define globally somewhere and then use all over the code base. For now, won't do it for these two instances.

**Comment by AmeerHajAli** [comment]
@ericl , this looks good to me.

**Comment by dmatch01** [comment]
@ericl Just wanted to give a gentle nudge for the merge.  Would like to continue with my testing once this merge is available.  Thank you! cc: @AmeerHajAli @DmitriGekhtman

**Review by ericl — APPROVED** [review]

**Comment by dmatch01** [comment]
@ericl Thank you!

**Pull request merged**
